contracts/crashpunks-v2-friedger.clar

```diff
@@ -0,0 +1,345 @@
+        (nft-transfer? crashpunks-v2 nftIndex owner recipient)
+    )
+)
+
+;; operable
+(define-read-only (is-approved (nftIndex uint) (operator principal))
+    (let ((owner (unwrap! (nft-get-owner? crashpunks-v2 nftIndex) ERR-COULDNT-GET-NFT-OWNER)))
+        (ok (is-owned-or-approved nftIndex operator owner))
+    )
+)
+
+;; operable
+(define-public (set-approved (nftIndex uint) (operator principal) (approved bool))
+    (let ((owner (unwrap! (nft-get-owner? crashpunks-v2 nftIndex) ERR-COULDNT-GET-NFT-OWNER)))
+        (ok (map-set approvals {owner: contract-caller, operator: operator, nft-index: nftIndex} approved))
+    )
+)
+
+(define-public (set-approved-all (operator principal) (approved bool))
+    (ok (map-set approvals-all {owner: contract-caller, operator: operator} approved))
+)
+
+;; public methods
+
+;; upgrade from v1 to v2
+;; Owner of crashpunks v1 calls this upgrade function
+;; 1. Transfers v1 NFT to this contract
+;; 2. This contract mints the v2 NFT with the same nftIndex for contract-caller
+;; 3. This contract burns the v1 NFT
+(define-public (upgrade-v1-to-v2 (nftIndex uint))
+    ;; TODO: MAKE SURE THESE CONTRACT CALLS WORK, MAKE SURE THE CONRACT ADDRESSES WORKS FOR MAINNET
+    (begin 
+        ;; assert contract caller owns the v1 NFT at this nftIndex
+        (asserts! (is-eq contract-caller (unwrap! (unwrap! (contract-call? .crashpunks-v1 get-owner nftIndex) ERR-NOT-V1-OWNER) ERR-NOT-V1-OWNER)) ERR-NOT-V1-OWNER)
+
+        ;; 1. transfer v1 NFT to this contract
+        (try! (contract-call? .crashpunks-v1 transfer nftIndex contract-caller (as-contract tx-sender)))
+        
+        ;; 2. Mint the v2 NFT with the same nftIndex for contract-caller
+        (try! (nft-mint? crashpunks-v2 nftIndex contract-caller))
+
+        ;; 3. Burn the v1 NFT
+        (try! (contract-call? .crashpunks-v1 burn nftIndex (as-contract tx-sender)))
+        (ok true)
+    )
+)
+
+(define-public (batch-upgrade-v1-to-v2 (entries (list 200 uint)))
+    (ok (fold upgrade-v1-to-v2-helper entries false))
+)
+
+(define-public (mint-token)
+    (let (
+            (mintCounter (var-get mint-counter))
+            (mintPassBalance (get-mint-pass-balance contract-caller))
+        )
+        (asserts! (< mintCounter COLLECTION-MAX-SUPPLY) ERR-COLLECTION-LIMIT-REACHED)
+        (asserts! (> mintPassBalance u0) ERR-MINT-PASS-LIMIT-REACHED)
+
+        (try! (paymint-split MINT-PRICE contract-caller))
+        (try! (nft-mint? crashpunks-v2 mintCounter contract-caller))
+        (var-set mint-counter (+ mintCounter u1))
+        (map-set mint-pass contract-caller (- mintPassBalance u1))
+        (ok true)
+    )
+)
+
+;; only size of list matters, content of list doesn't matter
+(define-public (batch-mint-token (entries (list 20 uint)))
+    (ok (fold mint-token-helper entries false))
+)
+
+;; fail-safe: allow admin to airdrop to recipient, hopefully will never be used
+(define-public (admin-mint-airdrop (recipient principal) (nftIndex uint))
```

Comment:
`admin-mint-airdrop` means that v1 is NOT burnt.

Reply:
yup, that's why it's a worst case scenario function. this is for when the v1 NFT is stuck and the owner is unable to use the upgrade function.